src/main/java/com/scalar/cassy/scheduler/CassyClient.java

```diff
@@ -0,0 +1,144 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class CassyClient {
+  private final CassyGrpc.CassyBlockingStub blockingStub;
+
+  public CassyClient() {
+    this(
+        ManagedChannelBuilder.forAddress(
+                System.getenv().getOrDefault("CASSY_SCHEDULER_HOST", "localhost"),
+                Integer.parseInt(System.getenv().getOrDefault("CASSY_SCHEDULER_PORT", "20051")))
+            .usePlaintext());
+  }
+
+  public CassyClient(ManagedChannelBuilder<?> channelBuilder) {
+    Channel channel = channelBuilder.build();
+    blockingStub = CassyGrpc.newBlockingStub(channel);
+  }
+
+  public void takeClusterSnapshot(String clusterId, int timeout) throws Exception {
+    BackupRequest backupRequest =
+        BackupRequest.newBuilder()
+            .setClusterId(clusterId)
+            .setBackupType(BackupType.CLUSTER_SNAPSHOT.get())
+            .build();
+    BackupResponse backupResponse = blockingStub.takeBackup(backupRequest);
+    startTask(backupResponse, timeout);
+  }
+
+  public void takeNodeSnapshot(String clusterId, int timeout, List<String> targetIps)
+      throws InterruptedException, TimeoutException, ExecutionException {
+    BackupRequest backupRequest =
+        BackupRequest.newBuilder()
+            .setClusterId(clusterId)
+            .setBackupType(BackupType.NODE_SNAPSHOT.get())
+            .build();
+    if (!targetIps.isEmpty()) {
+      backupRequest =
+          BackupRequest.newBuilder(backupRequest)
+              .addAllTargetIps(targetIps)
+              .build();
+    }
+
+    startTask(blockingStub.takeBackup(backupRequest), timeout);
+  }
+
+  public void takeIncrementalBackup(String clusterId, int timeout, List<String> targetIps)
+      throws InterruptedException, TimeoutException, ExecutionException {
+
+    // get cluster target ips if no target ips are set
+    if (targetIps.isEmpty()) {
+      targetIps =
+          blockingStub.listClusters(
+              ClusterListingRequest.newBuilder().setClusterId(clusterId).build()).getEntries(0).getTargetIpsList();
+    }
+
+    // for each target_ip, get a backup listing response
+    for (String targetIp : targetIps) {
+      BackupListingResponse backupListingResponse =
+          blockingStub.listBackups(
+              BackupListingRequest.newBuilder()
+                  .setClusterId(clusterId)
+                  .setTargetIp(targetIp)
+                  .build());
+      List<BackupListingResponse.Entry> entries =
+          backupListingResponse.getEntriesList().stream()
+              .filter(e -> e.getBackupType() != BackupType.NODE_INCREMENTAL.get())
+              .collect(Collectors.toList());
+      // get the most recent snapshot
+      if (entries.get(0).getStatusValue() != OperationStatus.COMPLETED_VALUE) {
+        // if snapshot is not COMPLETED, throw an exception
+        throw new BackupException(
```

Comment:
I think it should be either
* Take whatever incremental backups that can be taken
* Take if all the incremental backups can be taken

But the current algorithm is not either of them because it throws an exception when it finds some error in the most recent snapshots.

Maybe `Take whatever incremental backups that can be taken` is fine because taking snapshots works similary.

Reply:
That make sense. I'll update the logic.